src/object/obj_layout.c

```diff
@@ -58,18 +58,32 @@ obj_layout_diff(struct pl_map *map, daos_unit_oid_t oid, uint32_t new_ver, uint3
 	if (rc)
 		D_GOTO(out, rc);
 
-	if (new_layout->ol_shards[shard].po_target != old_layout->ol_shards[shard].po_target) {
-		*tgt = new_layout->ol_shards[shard].po_target;
-		*shard_p = shard;
-		D_GOTO(out, rc = 1);
-	}
-
 	/* If the new layout changes dkey placement, i.e. dkey->grp, dkey->ec_start changes,
 	 * then all shards needs to be changed.
 	 */
 	if (new_ver == 1 && daos_obj_id_is_ec(oid.id_pub)) {
-		*tgt = new_layout->ol_shards[shard].po_target;
-		*shard_p = shard;
+		struct daos_oclass_attr	*oc;
+		unsigned int	grp_size;
+		unsigned int	grp_start;
+		int		i;
+
+		oc = daos_oclass_attr_find(oid.id_pub, NULL);
+		D_ASSERT(oc != NULL);
+		grp_size = daos_oclass_grp_size(oc);
+
+		D_ASSERT(grp_size < array_size);
+		grp_start = (shard / grp_size) * grp_size;
+		for (i = 0; i < grp_size; i++) {
+			tgts[i] = new_layout->ol_shards[grp_start + i].po_target;
+			shards[i] = grp_start + i;
```

Comment:
if (new_layout->ol_shards[shards[i]].po_target == old_layout->ol_shards[shards[i]].po_target)
for this case seems need not migrate? seems that is possible?

Reply:
if ec parity rotation is enabled, then it needs to migrate for upgrading, so within the group, dkey might be in different shard,  because parity shard has been changed.

Reply:
ok, i see, thanks.